src/tests/ftest/cart/util/daos_sys_logscan.py

```diff
@@ -0,0 +1,561 @@
+
+    # pool map version update events
+    upd_re = r"ds_pool_tgt_map_update.*([0-9a-fA-F]{8}): updated.*map: version=(\d+)->(\d+) pointer"
+    re_pmap_update = re.compile(upd_re)
+
+    # uniform rebuild string identifier rb=<pool_uuid>/<rebuild_ver>/<rebuild_gen>/<opcode_string>
+    rbid_re = r"rb=([0-9a-fA-F]{8})/(\d+)/(\d+)/(\w+)"
+
+    # Future possibility: match the rebuild preliminary steps
+    # rebuild_task_ult() wait for scheduling, and map dist - both would info to match on.
+    # re.compile(r"rebuild_task_ult\(\).*rebuild task sleep (\d+) second")
+    # re.compile(r"rebuild_task_ult\(\).*map_dist_ver (\d+) map ver (\d+)")
+
+    # Rebuild: PS leader engine starting and status checking a given operation
+    # statuses: "scanning", "pulling", "completed", "aborted"
+    ldr_start_re = "rebuild_leader_start.*" + rbid_re + "$"
+    ldr_status_re = r"rebuild_leader_status_check\(\).*" + rbid_re + r" \[(\w+)\].*duration=(\d+)"
+    re_rebuild_ldr_start = re.compile(ldr_start_re)
+    re_rebuild_ldr_status = re.compile(ldr_status_re)
+
+    # Legacy rebuild PS leader logging (before uniform rebuild string)
+    old_ldr_start_re = r"rebuild_leader_start.*([0-9a-fA-F]{8}).*version=(\d+)/(\d+).*op=(\w+)"
+    old_ldr_status_re = (r"rebuild_leader_status_check\(\) (\w+) \[(\w+)\] \(pool ([0-9a-fA-F]{8}) "
+                         r"leader (\d+) term (\d+).*ver=(\d+),gen (\d+).*duration=(\d+) secs")
+    re_old_ldr_start = re.compile(old_ldr_start_re)
+    re_old_ldr_status = re.compile(old_ldr_status_re)
+
+    def __init__(self):
+        # dictionaries indexed by pool UUID
+        self._pools = {}
+        self._highest_pmapver = {}
+        self._cur_ldr_rank = {}
+        self._cur_ldr_pid = {}
+        self._cur_term = {}
+
+        # filename to rank map
+        self._file_to_rank = {}
+
+        self._warnings = []
+        self._check_rb_new_fmt = True
+        self._check_rb_legacy_fmt = True
+        self._debug = False
+
+    def _warn(self, wmsg, fname, line=None):
+        full_msg = f"WARN file={fname}"
+        if line:
+            full_msg += f", line={line.lineno}"
+        full_msg += f": {wmsg}"
+
+        self._warnings.append(full_msg)
+        print(full_msg)
+
+    def get_warnings(self):
+        """Return all warnings stored when scanning engine log files"""
+        return self._warnings
+
+    def _set_rank(self, log_iter):
+        print(f"INFO: searching for rank in file {log_iter.fname}")
+        found = False
+        for line in log_iter.new_iter():
+            # when a rank assignment log line found (engine start)
+            match = self.re_rank_assign.match(line.get_msg())
+            if match:
+                self._file_to_rank[log_iter.fname] = int(match.group(1))
+                found = True
+                break
+
+            # Future enhancement: what about log rotation (not an engine start scenario)?
+        return found
```

Comment:
Nit - I'd recommend immediate return when possible
```suggestion
    def _set_rank(self, log_iter):
        print(f"INFO: searching for rank in file {log_iter.fname}")
        for line in log_iter.new_iter():
            # when a rank assignment log line found (engine start)
            match = self.re_rank_assign.match(line.get_msg())
            if match:
                self._file_to_rank[log_iter.fname] = int(match.group(1))
                return True

            # Future enhancement: what about log rotation (not an engine start scenario)?
        return False
```

Reply:
done in  latest commit